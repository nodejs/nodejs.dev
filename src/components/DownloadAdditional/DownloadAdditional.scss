$downloadable-item-height: 40px;

.download-additional {
  margin-top: 133px;
  width: 100%;
  font-weight: var(--font-weight-semibold);

  &__header {
    display: flex;
    justify-content: space-between;

    .download-toogle {
      margin-top: 0;
    }
  }

  &__title {
    font-size: var(--font-size-display5);
    line-height: var(--line-height-display4);
  }

  &__body {
    margin-top: 24px;
  }

  &__item {
    display: flex;
    flex-direction: column;
    border: 1px solid #d7d8df;
    outline: none;
    margin-top: -1px; // Cancels overlapping borders extra thickness
    padding: 16px 0 0 16px;
    height: $downloadable-item-height;
    transition: height 30ms ease-in;

    &__expanded {
      height: auto;
    }

    &__header {
      display: flex;
      flex-direction: row;
      &__arrow {
        margin-right: 16px;
        width: 0;
        height: 0;
        &__right {
          margin-top: 6px;
          border-top: 5px solid transparent;
          border-left: 6px solid #757575;
          border-bottom: 5px solid transparent;
          border-radius: 3px;
        }
        &__down {
          margin-top: 8px;
          border-left: 5px solid transparent;
          border-top: 6px solid #757575;
          border-right: 5px solid transparent;
          border-radius: 3px;
        }
      }
    }
    &__body {
      display: flex;
      flex-wrap: wrap;
      &__link {
        display: flex;
        align-items: center;
        background: var(--brand2);
        color: var(--brand6);
        text-decoration: none;
        font-size: 14px;
        line-height: 36px;
        border-radius: 40px;
        margin: 5px;
        padding: 0 15px 0 0;
      }
    }
  }
}
