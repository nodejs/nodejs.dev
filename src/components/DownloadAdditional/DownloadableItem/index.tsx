import React, { FC } from 'react';
import classnames from 'classnames';
import { Downloadable } from './downloadItems';
import styles from './index.module.scss';

interface DownloadButtonProps {
  className: string;
  link: string;
  label: string;
}

const DownloadButton: FC<DownloadButtonProps> = ({
  className,
  link,
  label,
}) => (
  <a className={className} href={link}>
    <i className="material-icons">get_app</i>
    {label}
  </a>
);

interface DownloadableItemProps {
  item: Downloadable;
  isExpanded: boolean;
  setExpandedItem: (itemName: string) => void;
}

const DownloadableItem: FC<DownloadableItemProps> = ({
  item,
  isExpanded,
  setExpandedItem,
}) => {
  const onClick = (): void =>
    isExpanded ? setExpandedItem('') : setExpandedItem(item.name);

  const itemClasses = classnames(styles.downloadableItem, {
    [styles.downloadableItemExpanded]: isExpanded,
  });

  const arrowClasses = isExpanded
    ? styles.headerArrowDown
    : styles.headerArrowRight;

  return (
    <div
      role="button"
      tabIndex={0}
      className={itemClasses}
      onKeyUp={onClick}
      onClick={onClick}
    >
      <div className={styles.header}>
        <div className={styles.headerArrow}>
          <div className={arrowClasses} />
        </div>
        {item.name}
      </div>

      {isExpanded && (
        <div className={styles.body}>
          {item.links.map(link => (
            <DownloadButton
              className={styles.bodyLink}
              key={link.label}
              link={link.path}
              label={link.label}
            />
          ))}
        </div>
      )}
    </div>
  );
};

export default DownloadableItem;
