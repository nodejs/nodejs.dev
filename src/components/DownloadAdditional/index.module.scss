.download-additional {
  font-weight: var(--font-weight-semibold);
  margin-top: 133px;
  width: 100%;

  &__header {
    align-items: center;
    display: flex;
    justify-content: space-between;

    @media (max-width: 800px) {
      align-items: center;
      flex-direction: column;
    }

    div {
      margin-top: 0;
    }
  }

  &__title {
    margin: 0;

    @media (max-width: 800px) {
      padding-bottom: var(--space-12);
    }
  }

  &__body {
    margin-top: 24px;
  }
}
