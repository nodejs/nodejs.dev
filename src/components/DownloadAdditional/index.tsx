import React, { useState, FC } from 'react';
import DownloadToggle from '../DownloadToggle';
import { ReleaseData } from '../../hooks/useReleaseHistory';
import { getDownloadableItemsList, Downloadable } from './downloadItems';
import './DownloadAdditional.scss';

const CLASS_NAME = 'download-additional';

interface DownloadButtonProps {
  className: string;
  link: string;
  label: string;
}

const DownloadButton: FC<DownloadButtonProps> = ({
  className,
  link,
  label,
}: DownloadButtonProps) => (
  <a className={className} href={link}>
    <i className="material-icons">get_app</i>
    {label}
  </a>
);

interface DownloadableItemProps {
  item: Downloadable;
  isExpanded: boolean;
  setExpandedItem: (itemName: string) => void;
}

const DownloadableItem: FC<DownloadableItemProps> = ({
  item,
  isExpanded,
  setExpandedItem,
}: DownloadableItemProps) => {
  const onClick = (): void =>
    isExpanded ? setExpandedItem('') : setExpandedItem(item.name);
  const classes = `${CLASS_NAME}__item`.concat(
    isExpanded ? ` ${CLASS_NAME}__item__expanded` : ''
  );
  return (
    <div
      role="button"
      tabIndex={0}
      className={classes}
      onKeyUp={onClick}
      onClick={onClick}
    >
      <div className={`${CLASS_NAME}__item__header`}>
        <div className={`${CLASS_NAME}__item__header__arrow`}>
          {isExpanded ? (
            <div className={`${CLASS_NAME}__item__header__arrow__down`} />
          ) : (
            <div className={`${CLASS_NAME}__item__header__arrow__right`} />
          )}
        </div>
        {item.name}
      </div>

      {isExpanded && (
        <div className={`${CLASS_NAME}__item__body`}>
          {item.links.map(link => (
            <DownloadButton
              className={`${CLASS_NAME}__item__body__link`}
              key={link.label}
              link={link.path}
              label={link.label}
            />
          ))}
        </div>
      )}
    </div>
  );
};

interface DownloadAdditionalProps {
  line?: ReleaseData;
  selectedTypeRelease: string;
  handleTypeReleaseToggle: (selected: React.SetStateAction<string>) => void;
}

const DownloadAdditional: FC<DownloadAdditionalProps> = ({
  line,
  selectedTypeRelease,
  handleTypeReleaseToggle,
}: DownloadAdditionalProps): JSX.Element => {
  const [expandedItem, setExpandedItem] = useState('');
  return (
    <div className={CLASS_NAME}>
      <div className={`${CLASS_NAME}__header`}>
        <h3 className={`${CLASS_NAME}__title`}>Additional Downloads</h3>
        <DownloadToggle
          selected={selectedTypeRelease}
          handleClick={handleTypeReleaseToggle}
          showDescription={false}
        />
      </div>
      <div className={`${CLASS_NAME}__body`}>
        {line &&
          getDownloadableItemsList(line?.version).map(item => (
            <DownloadableItem
              key={item.name}
              item={item}
              isExpanded={expandedItem === item.name}
              setExpandedItem={setExpandedItem}
            />
          ))}
      </div>
    </div>
  );
};

export default DownloadAdditional;
