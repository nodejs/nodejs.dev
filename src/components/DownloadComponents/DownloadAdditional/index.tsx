import React, { useState } from 'react';
import { FormattedMessage } from 'react-intl';
import DownloadableItem from './DownloadableItem';
import DownloadToggle from '../DownloadToggle';
import { getDownloadableItemsList } from './DownloadableItem/downloadItems';
import { NodeReleaseData } from '../../../types';
import styles from './index.module.scss';

interface DownloadAdditionalProps {
  line?: NodeReleaseData;
  selectedTypeRelease: string;
  handleTypeReleaseToggle: (selected: React.SetStateAction<string>) => void;
}

const DownloadAdditional = ({
  line,
  selectedTypeRelease,
  handleTypeReleaseToggle,
}: DownloadAdditionalProps) => {
  const [expandedItem, setExpandedItem] = useState('');

  return (
    <div className={styles.downloadAdditional}>
      <div className={styles.header}>
        <h3 className={styles.title}>
          <FormattedMessage id="components.downloadAdditional.title" />
        </h3>
        <DownloadToggle
          selected={selectedTypeRelease}
          handleClick={handleTypeReleaseToggle}
          showDescription={false}
        />
      </div>
      <div className={styles.body}>
        {line &&
          getDownloadableItemsList(line.fullVersion).map(item => (
            <DownloadableItem
              key={item.name}
              item={item}
              isExpanded={expandedItem === item.name}
              setExpandedItem={setExpandedItem}
            />
          ))}
      </div>
    </div>
  );
};

export default DownloadAdditional;
