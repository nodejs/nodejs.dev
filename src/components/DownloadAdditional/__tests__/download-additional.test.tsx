import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import DownloadAdditional from '..';

describe('DownloadAdditional component', (): void => {
  const releaseData = {
    date: 'date',
    version: 'version',
    files: [],
    lts: true,
    v8: 'v8',
  };

  it('renders correctly', (): void => {
    const { container } = render(
      <DownloadAdditional
        selectedTypeRelease="LTS"
        handleTypeReleaseToggle={(): null => null}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with release data', (): void => {
    const { container } = render(
      <DownloadAdditional
        line={releaseData}
        selectedTypeRelease="LTS"
        handleTypeReleaseToggle={(): null => null}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('allows to expand items', () => {
    const { container } = render(
      <DownloadAdditional
        line={releaseData}
        selectedTypeRelease="LTS"
        handleTypeReleaseToggle={(): null => null}
      />
    );

    const downloadItem = container.querySelector(
      '.download-additional__item'
    ) as Element;

    // expand installers list
    fireEvent.click(downloadItem);

    expect(container).toMatchSnapshot();
  });

  it('allows to collapse expanded items', () => {
    const { container } = render(
      <DownloadAdditional
        line={releaseData}
        selectedTypeRelease="LTS"
        handleTypeReleaseToggle={(): null => null}
      />
    );

    const downloadItem = container.querySelector(
      '.download-additional__item'
    ) as Element;

    // expand installers list
    fireEvent.click(downloadItem);
    // collapse it back to check all relevant html classes was set
    fireEvent.click(downloadItem);

    expect(container).toMatchSnapshot();
  });
});
