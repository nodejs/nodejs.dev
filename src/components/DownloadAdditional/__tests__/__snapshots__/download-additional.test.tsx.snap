// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadAdditional component renders correctly 1`] = `
<div>
  <div
    class="download-additional"
  >
    <div
      class="download-additional__header"
    >
      <h3
        class="download-additional__title"
      >
        Additional Downloads
      </h3>
      <div
        class="download-toogle"
      >
        <div
          class="download-toogle__selector"
        >
          <div
            class="download-toogle__switch"
          >
            <button
              aria-checked="true"
              aria-label="Show LTS versions"
              class="download-toogle__button -active"
              role="switch"
              type="button"
            >
              LTS
            </button>
            <button
              aria-checked="true"
              aria-label="Show LTS versions"
              class="download-toogle__button -current"
              role="switch"
              type="button"
            >
              Current
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="download-additional__body"
    />
  </div>
</div>
`;
